### PR TITLE
Added scheduler as arg to Observatory and error NoScheduler

### PR DESCRIPTION
**Pull request by ImmortalRabbit**
I added a new argument to Observatory class and  a new error NoScheduler. 

## Description

Currently, I just don't use _create_scheduler. Functions that rely on scheduler can throw an error NoScheduler("Scheduler not found.") if scheduler is None. I am not quite sure that I did it right. I literally wrote code in every function that checks function that rely on scheduler. Is that okay? Also, I have a question, according to my understanding I will have to create a function `create_scheduler_from_config(config = None, logger=None, observer=None)`. Code of this function will be similar to `_create_scheduler`. However, if I will decide to test it, what kind of observer should I send to it? In Observatory class observer changes multiple times and I am not quite sure what kind of observer it should receive. I am confused at this part.

## Related Issue

#837 #836 #838 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

@wtgee @jamessynge

### CODE REVIEW

**Comment by wtgee** [comment]
> I literally wrote code in every function that checks function that rely on scheduler. Is that okay?

I think we will inevitably need to do this so it's the right direction. However, in order to properly observe we will need more items than just a scheduler, i.e. we need a camera, a mount, etc.

What we should probably do is create another method (it could potentially also be a property), `can_observe` that can check all of these items (as well as the weather safety) and then replace what you have done here with a call to `can_observe`.  For now you can just put the check for scheduler in `can_observe` but eventually it will have a few other items in there as well.

**Comment by ImmortalRabbit** [comment]
@wtgee @jamessynge What do say about my new commits?

Is that correct pull request. I just didn't know how to do divide scheduler process into separate pull request because this codes are connected, so I decided to do a few commits with one pull request for each of them (scheduler, mount, dome and writing tests). 

Currently, I just implemented create_scheduler_from_config and in order to check it I create scheduler in Observatory class. 

Next step will be moving calling create_scheduler_from_config outside of Observatory class.

File: pocs/observatory.py [hunk 1]
```diff
@@ -59,8 +58,10 @@ def __init__(self, cameras=None, *args, **kwargs):
         self.dome = pocs.dome.create_dome_from_config(self.config, logger=self.logger)
 
         self.logger.info('\tSetting up scheduler')
-        self.scheduler = None
-        self._create_scheduler()
+        self.scheduler = scheduler
+        if self.scheduler is None:
+            self.scheduler = pocs.scheduler.create_scheduler_from_config(self.config, logger=self.logger,
```
Comment (wtgee) [inline pocs/observatory.py]: We don't want to automatically create a scheduler if one is not passed during construction. Probably nothing needs to happen here, except maybe a log message stating that there is no scheduler.

File: pocs/observatory.py [hunk 2]
```diff
@@ -130,17 +131,23 @@ def primary_camera(self, cam):
 
     @property
     def current_observation(self):
+        if not self.can_observe:
+            raise error.NoObservation("No valid observations found.")
```
Comment (wtgee) [inline pocs/observatory.py]: `can_observe` could evaluate to False for a number of reasons (no scheduler, no cameras, no mount). That is not the same thing as saying there is not a valid observation. This message implies that all of the hardware is correct and that the scheduler is working, it's just that there are no particular targets available at this time, which is not what we want to say.

The specific reason that `can_observe` is False should be handled inside `can_observe` itself.  

Also, one could technically have a `scheduler`, and a `current_observation`, and still `can_observe` is False, e.g. when there are no cameras present.  We shouldn't be prevented from getting the `current_observation` in that situation.

File: pocs/observatory.py [hunk 3]
```diff
@@ -130,17 +131,23 @@ def primary_camera(self, cam):
...
         return self.scheduler.current_observation
 
     @current_observation.setter
     def current_observation(self, new_observation):
+        if not self.can_observe:
+            raise error.NoObservation("No valid observations found.")
```
Comment (wtgee) [inline pocs/observatory.py]: Same as above.

File: pocs/observatory.py [hunk 4]
```diff
@@ -130,17 +131,23 @@ def primary_camera(self, cam):
...
         self.scheduler.current_observation = new_observation
 
     @property
     def has_dome(self):
         return self.dome is not None
 
-
+    @property
+    def can_observe(self):
+        return self.scheduler is not None
```
Comment (wtgee) [inline pocs/observatory.py]: This is going to expand to include things like the camera, mount, etc. Let's set up the framework to test each of those and give appropriate message if not found.

This will end up something like:
```python
def can_observe(self):
   can_observe = True

    if can_observe and self.scheduler is None:
        self.logger.info(f'Scheduler not present, cannot observe')
        can_observe = False

    # Same thing for mount
    # Same thing for cameras
    # Etc.

    return can_observe
```

Also, docstrings. ;)

File: pocs/observatory.py [hunk 5]
```diff
@@ -202,6 +209,9 @@ def status(self):
         """
         status = {}
         try:
+            if not self.can_observe:
```
Comment (wtgee) [inline pocs/observatory.py]: We should not be raising an Exception in the status call. We always want to be able to return a status. That status should indicate if observing is possible or not. Basically, `status['can_observe'] = self.can_observe` is all that's needed here.

File: pocs/observatory.py [hunk 6]
```diff
@@ -255,6 +265,9 @@ def get_observation(self, *args, **kwargs):
 
         self.logger.debug("Getting observation for observatory")
 
+        if not self.can_observe:
+            raise error.NoObservation("No valid observations found.")
```
Comment (wtgee) [inline pocs/observatory.py]: Same comments as above w.r.t. the observations vs the scheduler.

File: pocs/observatory.py [hunk 7]
```diff
@@ -411,6 +427,9 @@ def analyze_recent(self):
             "Analyzing recent image using pointing image: '{}'".format(pointing_image))
 
         try:
+            if not self.can_observe:
```
Comment (wtgee) [inline pocs/observatory.py]: Not sure this is necessary here in the analyze face.

File: pocs/observatory.py [hunk 8]
```diff
@@ -500,6 +520,9 @@ def get_standard_headers(self, observation=None):
 
         self.logger.debug("Getting headers for : {}".format(observation))
 
+        if not self.can_observe:
```
Comment (wtgee) [inline pocs/observatory.py]: Not necessary here.

File: pocs/scheduler/__init__.py [hunk 9]
```diff
@@ -1 +1,69 @@
...
+
+    if not logger:
+        logger = logger_module.get_root_logger()
+
+    if 'scheduler' not in config:
+        logger.info("No scheduler in config")
+        return None
+
+    if not observer:
+        logger.info("No valid observation found.")
```
Comment (wtgee) [inline pocs/scheduler/__init__.py]:
```suggestion
        logger.info("No valid Observer found.")
```

An `Observer` is different from an `Observation`. The `Observer` is the telescope itself, which has knowledge of the timezone and location where it is sitting.

File: pocs/observatory.py (hunk 2 above)
Comment (ImmortalRabbit) [reply pocs/observatory.py]: According to your commentary, I understood that I should not raise error when can_observe is False. can_observe should display reason of it's falseness ( like no scheduler, no cameras, no mount) and should return True or False. Then, when I will start using it, I am thinking that I can do like that
```if not can_observe:
         return None #what else should I return instead of None or how else I have to stop function if it's not possible to observe```

File: pocs/scheduler/__init__.py [hunk 10]
```diff
@@ -1 +1,69 @@
...
+    """ Sets up the scheduler that will be used by the observatory """
+
+    if not logger:
+        logger = logger_module.get_root_logger()
+
+    if 'scheduler' not in config:
+        logger.info("No scheduler in config")
+        return None
+
+    if not observer:
```
[the remark this answers is not in the document]
Comment (wtgee) [reply pocs/scheduler/__init__.py]: @jamessynge any thoughts on this?

[the remark this answers is not in the document]
Comment (ImmortalRabbit) [reply pocs/scheduler/__init__.py]: > @jamessynge any thoughts on this?

@wtgee I made separate issue #840.

File: pocs/observatory.py [hunk 11]
```diff
@@ -303,6 +329,9 @@ def cleanup_observations(self, upload_images=None, make_timelapse=None, keep_jpg
             except KeyError:
                 keep_jpgs = True
 
+        if not self.can_observe:
```
[the remark this answers is not in the document]
Comment (ImmortalRabbit) [reply pocs/observatory.py]: > For now let's just check for `self.scheduler` here (again with no error) and not worry about the rest for this PR. However we will want to make an Issue to handle the cleanup situation when a scheduler is removed.

@wtgee I made separate issue #840.

**Comment by ImmortalRabbit** [comment]
> HI @ImmortalRabbit, there are still some of the `NoObservation`s popping up in a few places they probably shouldn't. Also it looks like the tests aren't passing although I haven't yet looked at the reasons why.

It's probably because _create_scheduler is not called right know and create_scheduler_from_config is not used yet. For example, test_bad_scheduler sends config with bad scheduler type which had to lead to error during creation of scheduler but it didn't because Observatory class have to receive scheduler and if it doesn't receives it sets scheduler as None. Therefore, Observatory class can be initialized with scheduler which is None.

File: pocs/tests/test_scheduler.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+from pocs.scheduler import create_scheduler_from_config
+from pocs.site_location_details import setup_site_location_details_from_config
+from pocs.utils import error
+
+
+def test_bad_scheduler_type(config):
+    conf = config.copy()
+    conf['scheduler']['type'] = 'foobar'
+    details = setup_site_location_details_from_config(config)
+    with pytest.raises(error.NotFound):
```
Comment (wtgee) [inline pocs/tests/test_scheduler.py]: This is actually a good example of how code coverage can help expose bugs even if the tests don't seem to. Looking at your PR I'm realizing we've had a problem in our scheduler testing, which is not something that you introduced but instead was already there.

If you look at the code coverage for the scheduler you'll see that there is one line that is not getting hit, namely the error that should be raised for an invalid scheduler type ([line 64](https://codecov.io/gh/panoptes/POCS/pull/838/diff?src=pr&el=tree#D5-64)). What should be happening is that the code should be unable to find the `foobar` scheduler (because it doesn't exist) and so it should raise a ` NotFound` exception on line 64. However the red is telling us that piece of code is never getting run, which means we are never actually using it.

What is going on is that this test, which should be checking for the bad scheduler type, is actually failing because of the missing fields path check on [line 37](https://github.com/panoptes/POCS/blob/70416431f23476a3b535513ff59ffc1057e6407c/pocs/scheduler/__init__.py#L37). (The fields file is checked in the next test below).  So, basically we are not really ever checking for a bad scheduler type.

And looking even deeper, the `load_module` function itself is [catching the ImportError](https://github.com/panoptes/POCS/blob/develop/pocs/utils/__init__.py#L316), and is raising the `error.NotFound`, which means the [ImportError on line 63](https://github.com/panoptes/POCS/blob/70416431f23476a3b535513ff59ffc1057e6407c/pocs/scheduler/__init__.py#L63) could never possibly get raised because `load_module` is re-raising the `ImportError` but as an `error.NotFound` instead.

Again, this was not something you have done but was something that was already there. However let's use this time to go ahead and clean it up.

Two things need to be done here:

* [Line 63](https://github.com/panoptes/POCS/blob/70416431f23476a3b535513ff59ffc1057e6407c/pocs/scheduler/__init__.py#L63) should change to be `except error.NotFound as e`. This is a little bit pointless as it just re-raises the same exception on the next line and so you could alternatively just remove the entire `try/except` block (lines 39 and 63). However let's keep it and just re-raise the `error.NotFound` for now. In the future we would maybe want to do something smarter there instead (such as create a `error.ImportError` of our own), so it makes sense to leave it in place.
* This test needs to have a valid `conf['scheduler']['fields_file']`. That is, it should be set to a real path name such that the existence check (line [37](https://github.com/panoptes/POCS/blob/70416431f23476a3b535513ff59ffc1057e6407c/pocs/scheduler/__init__.py#L37)) evaluates to True.

Once this is done the tests should pass, as they do now, but that [red marker on line 64](https://codecov.io/gh/panoptes/POCS/pull/838/diff?src=pr&el=tree#D5-64) should disappear and our coverage should go up (a tiny bit).

That's a pretty subtle bug. Let me know if anything doesn't make sense.  Thanks!

Comment (wtgee) [reply pocs/tests/test_scheduler.py]: But yes, this is a good example of why you should look at the coverage reports and try to figure out _why_ the red chunks are red.

Comment (ImmortalRabbit) [reply pocs/tests/test_scheduler.py]: @wtgee Should I try to increase coverage in Observatory class?

File: pocs/observatory.py [hunk 13]
```diff
@@ -490,6 +526,10 @@ def get_standard_headers(self, observation=None):
         Returns:
             dict: The standard headers
         """
+
+        if not self.can_observe:
```
Comment (wtgee) [inline pocs/observatory.py]: This check isn't necessary as should be handled by the `assert` on line 536 below.

File: pocs/tests/test_scheduler.py (hunk 12 above)
Comment (wtgee) [reply pocs/tests/test_scheduler.py]: Left a message in slack a few days ago, but no, not unless it is directly related to this PR. We want to try to keep PRs focused.

File: pocs/tests/test_observatory.py [hunk 14]
```diff
@@ -27,7 +29,10 @@ def simulator():
 @pytest.fixture
 def observatory(config, simulator, images_dir):
     """Return a valid Observatory instance with a specific config."""
+    details = setup_site_location_details_from_config(config)
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: See comments above but `details` is also a meaningless name as it could refer to the details of anything. This is `location_info` or `site_info`. Since it contains keys called `location` then it should probably be `site_details`.

File: pocs/site_location_details/__init__.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
+from astroplan import Observer
```
Comment (wtgee) [inline pocs/site_location_details/__init__.py]: This is an awkard name for a directory and also kind of just an awkward file. All of this is really just setting up configuration details and shouldn't really be set up as a module.

Could move this file into `pocs/utils/location.py` for now although we might want to change that in the long run.

File: pocs/tests/test_observatory.py [hunk 16]
```diff
@@ -1,16 +1,18 @@
...
-from pocs import hardware
 import pocs.version
+from pocs import hardware
+from pocs.camera import create_cameras_from_config
 from pocs.observatory import Observatory
+from pocs.scheduler import create_scheduler_from_config
 from pocs.scheduler.dispatch import Scheduler
 from pocs.scheduler.observation import Observation
-from pocs.camera import create_cameras_from_config
+from pocs.site_location_details import setup_site_location_details_from_config
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: See comments in file above.

File: pocs/tests/test_observatory.py [hunk 17]
```diff
@@ -37,6 +42,19 @@ def observatory(config, simulator, images_dir):
     return obs
 
 
+def test_camera_already_exists(observatory, config):
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: I see you added this already. We try to keep PRs focused on a single thing and adding tests for the camera doesn't really have anything to do with the scheduler. Ok to leave for now but in the future I'd probably ask you to more to a separate PR.

File: pocs/tests/test_observatory.py [hunk 18]
```diff
@@ -37,6 +42,19 @@ def observatory(config, simulator, images_dir):
...
+    cameras = create_cameras_from_config(config)
+    for cam_name, cam in cameras.items():
+        observatory.add_camera(cam_name, cam)
+
+
+def test_remove_cameras(observatory, config):
+    obs = observatory
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: This line is unnecessary.

File: pocs/tests/test_observatory.py [hunk 19]
```diff
@@ -69,20 +87,44 @@ def test_bad_mount_driver(config):
...
 
 
-def test_bad_scheduler(config):
+def test_can_observe_no_scheduler(config):
     conf = config.copy()
-    simulator = ['all']
-    conf['scheduler']['type'] = 'foobar'
-    with pytest.raises(error.NotFound):
-        Observatory(simulator=simulator, config=conf, ignore_local_config=True)
+    cameras = create_cameras_from_config(conf)
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: No reason to add the cameras to this test. The Observatory should be able to be created without cameras and you are trying to test the scheduler here, so you want to make a minimal example required to test the scheduler.

Comment (wtgee) [reply pocs/tests/test_observatory.py]: Also, you've removed the test for a badly named scheduler, which is different from no scheduler.

File: pocs/tests/test_observatory.py [hunk 20]
```diff
@@ -69,20 +87,44 @@ def test_bad_mount_driver(config):
...
+
+    obs = Observatory(cameras=cameras,
+                      scheduler=None,
+                      config=conf
+                      )
 
+    assert obs.can_observe is False
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: You've probably added the cameras because otherwise this is `False`. The situation for no scheduler is different from the situation with a badly named scheduler so that needs to be resolved first.

File: pocs/tests/test_observatory.py (hunk 19 above)
Comment (wtgee) [reply pocs/tests/test_observatory.py]: I guess you have that down in the test_scehduler, so that is okay.

Comment (wtgee) [reply pocs/tests/test_observatory.py]: You can capture the logs that are output and make sure it is outputting the correct log message rather than just checking for the False below.

File: pocs/tests/test_observatory.py [hunk 21]
```diff
@@ -69,20 +87,44 @@ def test_bad_mount_driver(config):
...
 
-def test_bad_scheduler_fields_file(config):
+
+def test_can_observe_no_cameras(config):
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: This test and the next test should follow above. That is, we should be testing for the correct log message with minimal examples added. You will want to do that for the cameras since you created the `can_observe` method and the cameras are already part of that. However see not on mount below.

File: pocs/tests/test_observatory.py [hunk 22]
```diff
@@ -69,20 +87,44 @@ def test_bad_mount_driver(config):
...
+    obs = Observatory(cameras=None,
+                      scheduler=scheduler,
+                      config=conf
+                      )
+
+    assert obs.can_observe is False
+
+
+def test_can_observe_no_mount(config):
+    conf = config.copy()
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: We haven't yet moved the mount creation out of the Observatory so probably didn't need to modify the `can_observe` at this point, which means we don't need to create this test right now.

File: pocs/tests/test_observatory.py [hunk 23]
```diff
@@ -122,6 +164,12 @@ def test_primary_camera(observatory):
     assert observatory.primary_camera is not None
 
 
+def test_primary_camera_no_primary_camera(observatory):
+    obs = observatory
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: This line is not necessary.

Did this test increase coverage of some line? I would be surprised if we weren't already testing that.

File: pocs/tests/test_observatory.py [hunk 24]
```diff
@@ -209,6 +257,13 @@ def test_standard_headers(observatory):
     assert headers['longitude'] == test_headers['longitude']
 
 
+def test_get_standard_headers_no_scheduler(observatory):
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: We *do* want to be able to get headers without a scheduler, as mentioned in the `get_standard_headers` above.

File: pocs/tests/test_observatory.py [hunk 25]
```diff
@@ -227,6 +282,23 @@ def test_get_observation(observatory):
     assert observatory.current_observation == observation
 
 
+def test_get_observation_no_scheduler(observatory):
+    obs = observatory
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: This line is unnecessary.

File: pocs/tests/test_observatory.py [hunk 26]
```diff
@@ -227,6 +282,23 @@ def test_get_observation(observatory):
...
+    obs.scheduler = None
+    assert obs.get_observation() is None
+
+
+def test_cleanup_observations_no_scheduler(observatory):
+    obs = observatory
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: Same.

File: pocs/tests/test_observatory.py [hunk 27]
```diff
@@ -227,6 +282,23 @@ def test_get_observation(observatory):
...
+    obs.scheduler = None
+    assert obs.cleanup_observations() is None
+
+
+def test_cleanup_observations_keep_jpgs(observatory):
+    obs = observatory
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: Same.

File: pocs/tests/test_observatory.py [hunk 28]
```diff
@@ -227,6 +282,23 @@ def test_get_observation(observatory):
...
+    assert obs.get_observation() is None
+
+
+def test_cleanup_observations_no_scheduler(observatory):
+    obs = observatory
+    obs.scheduler = None
+    assert obs.cleanup_observations() is None
+
+
+def test_cleanup_observations_keep_jpgs(observatory):
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]: This is unrelated to this PR and I'm not sure is testing what we want. Let's just remove for now.

File: pocs/tests/test_pocs.py [hunk 29]
```diff
@@ -1,18 +1,20 @@
...
 
+import pytest
 from astropy import units as u
 
 from pocs import hardware
+from pocs.camera import create_cameras_from_config
 from pocs.core import POCS
 from pocs.observatory import Observatory
+from pocs.scheduler import create_scheduler_from_config
+from pocs.site_location_details import setup_site_location_details_from_config
```
Comment (wtgee) [inline pocs/tests/test_pocs.py]: See note about this file.

File: pocs/tests/test_observatory.py (hunk 23 above)
Comment (ImmortalRabbit) [reply pocs/tests/test_observatory.py]: > This line is not necessary.
> 
> Did this test increase coverage of some line? I would be surprised if we weren't already testing that.

I wrote this test because it wasn't tested. It was marked in red color.

File: pocs/tests/test_observatory.py [hunk 30]
```diff
@@ -1,17 +1,19 @@
...
 import pocs.version
+from pocs import hardware
+from pocs.camera import create_cameras_from_config
 from pocs.observatory import Observatory
+from pocs.scheduler import create_scheduler_from_config
 from pocs.scheduler.dispatch import Scheduler
 from pocs.scheduler.observation import Observation
-from pocs.camera import create_cameras_from_config
 from pocs.utils import error
+from pocs.utils.location import setup_site_location_details_from_config
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]:
```suggestion
from pocs.utils.location import create_location_from_config
```

Let's just call this `create_location_from_config` even though it returns more than just a strict `location`.

File: pocs/tests/test_observatory.py [hunk 31]
```diff
@@ -27,7 +29,10 @@ def simulator():
 @pytest.fixture
 def observatory(config, simulator, images_dir):
     """Return a valid Observatory instance with a specific config."""
+    site_details = setup_site_location_details_from_config(config)
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]:
```suggestion
    site_details = create_location_from_config(config)
```

File: pocs/tests/test_observatory.py [hunk 32]
```diff
@@ -69,20 +86,17 @@ def test_bad_mount_driver(config):
...
-    conf = config.copy()
-    simulator = ['all']
-    conf['scheduler']['fields_file'] = 'foobar'
-    with pytest.raises(error.NotFound):
-        Observatory(simulator=simulator, config=conf, ignore_local_config=True)
+    obs = Observatory(config=conf)
+    assert obs.can_observe is False
+    assert caplog.records[-1].levelname == "INFO" and caplog.records[
+        -1].message == "Scheduler not present, cannot observe."
+    site_details = setup_site_location_details_from_config(conf)
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]:
```suggestion
    site_details = create_location_from_config(conf)
```

File: pocs/tests/test_observatory.py [hunk 33]
```diff
@@ -147,8 +166,7 @@ def test_default_config(observatory):
     """ Creates a default Observatory and tests some of the basic parameters """
 
     assert observatory.location is not None
-    assert observatory.location.get('elevation') - \
-        observatory.config['location']['elevation'] < 1. * u.meter
+    assert observatory.location.get('elevation') - observatory.config['location']['elevation'] < 1. * u.meter
```
Comment (wtgee) [inline pocs/tests/test_observatory.py]:
```suggestion
    assert observatory.location.get('elevation') == pytest.approx(observatory.config['location']['elevation'], rel=1 * u.meter)
```

I'm not actually sure if this will work with the units in place. Might need to just add `.value` to the end of each.

But the `pytest.approx` is the better way to test things like this.

File: pocs/tests/test_pocs.py [hunk 34]
```diff
@@ -1,18 +1,20 @@
...
 from pocs.core import POCS
 from pocs.observatory import Observatory
+from pocs.scheduler import create_scheduler_from_config
 from pocs.utils import CountdownTimer
-from pocs.utils.messaging import PanMessaging
-from pocs.utils import error
 from pocs.utils import current_time
-from pocs.camera import create_cameras_from_config
+from pocs.utils import error
+from pocs.utils.location import setup_site_location_details_from_config
```
Comment (wtgee) [inline pocs/tests/test_pocs.py]:
```suggestion
from pocs.utils.location import create_location_from_config
```

File: pocs/tests/test_pocs.py [hunk 35]
```diff
@@ -42,10 +44,17 @@ def cameras(config):
 
 
 @pytest.fixture(scope='function')
-def observatory(config, db_type, cameras):
+def scheduler(config):
+    site_details = setup_site_location_details_from_config(config)
```
Comment (wtgee) [inline pocs/tests/test_pocs.py]:
```suggestion
    site_details = create_location_from_config(config)
```

File: pocs/tests/test_scheduler.py [hunk 36]
```diff
@@ -0,0 +1,32 @@
+import pytest
+
+from pocs.scheduler import create_scheduler_from_config
+from pocs.utils import error
+from pocs.utils.location import setup_site_location_details_from_config
```
Comment (wtgee) [inline pocs/tests/test_scheduler.py]:
```suggestion
from pocs.utils.location import create_location_from_config
```

File: pocs/tests/test_scheduler.py [hunk 37]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def test_bad_scheduler_type(config):
+    conf = config.copy()
+    conf['scheduler']['type'] = 'foobar'
+    site_details = setup_site_location_details_from_config(config)
```
Comment (wtgee) [inline pocs/tests/test_scheduler.py]:
```suggestion
    site_details = create_location_from_config(config)
```

File: pocs/tests/test_scheduler.py [hunk 38]
```diff
@@ -0,0 +1,32 @@
...
+    with pytest.raises(error.NotFound):
+        create_scheduler_from_config(conf, observer=site_details['observer'])
+
+
+def test_bad_scheduler_fields_file(config):
+    conf = config.copy()
+    conf['scheduler']['fields_file'] = 'foobar'
+    site_details = setup_site_location_details_from_config(config)
```
Comment (wtgee) [inline pocs/tests/test_scheduler.py]:
```suggestion
    site_details = create_location_from_config(config)
```

File: pocs/tests/test_scheduler.py [hunk 39]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def test_no_observer(config):
+    assert create_scheduler_from_config(config, observer=None) is None
+
+
+def test_no_scheduler_in_config(config):
+    conf = config.copy()
+    del conf['scheduler']
+    site_details = setup_site_location_details_from_config(conf)
```
Comment (wtgee) [inline pocs/tests/test_scheduler.py]:
```suggestion
    site_details = create_location_from_config(conf)
```

File: pocs/utils/location.py [hunk 40]
```diff
@@ -0,0 +1,75 @@
+from astroplan import Observer
+from astropy import units as u
+from astropy.coordinates import EarthLocation
+
+from pocs.utils import error
+from pocs.utils.logger import get_root_logger
+
+
+def setup_site_location_details_from_config(config, logger=None):
```
Comment (wtgee) [inline pocs/utils/location.py]:
```suggestion
def create_location_from_config(config, logger=None):
```

File: pocs/utils/location.py [hunk 41]
```diff
@@ -0,0 +1,75 @@
...
+    """
+     Sets up the site and location details
```
Comment (wtgee) [inline pocs/utils/location.py]:
```suggestion
     Sets up the site and location details.
```

Don't forget the periods! ;) I miss this all the time too.
